Adds Animation Frame De-Duplication

LemonInTheDark opened the pull request:
One of the big heartaches for wallening (among many) is animated wall icons
When they're too large, we get massive lag when going near a wall. This is particularly bad when there's a lot of frames
There are some cases where they could be split off into overlays, applied seperately. This doesn't apply to all cases though

So I was thinking, there's surely a lot of duplicate frames here, at least for the major suck (plasma walls, currently 2.7 mbs) 
You're busy with uh, I think a UI, so I decided to take a crack at it myself

I've written a function that walks the frames of an icon state, and simply removes the duplicates, extending the animation time 
This isn't the fastest possible implementation, both because this is my first time writing rust, and because I'm not like, doing it before insertion

It's more general tho, and lets us apply the proc to all the uses of IconState in the codebase

Sorry if uh, it's crap. Doin what I can

Oh right nearly forgot, this drops plasma walls down more then a megabyte, to 1.2mbs. Doesn't really do anything to heirophont walls, the other use of animations we have, but that's alright, I can make those into overlays anyhow

LemonInTheDark committed:
Adds Animation Frame De-Duplication

One of the big heartaches for wallening (among many) is animated wall
icons
When they're too large, we get massive lag when going near a wall. This
is particularly bad when there's a lot of frames
There are some cases where they could be split off into overlays,
applied seperately. This doesn't apply to all cases though

So I was thinking, there's surely a lot of duplicate frames here, at
least for the major suck (plasma walls, currently 2.7 mbs)
Action is busy with uh, I think a UI, so I decided to take a crack at it
myself

I've written a function that walks the frames of an icon state, and
simply removes the duplicates, extending the animation time
This isn't the fastest possible implementation, both because this is my
first time writing rust, and because I'm not like, doing it before
insertion

It's more general tho, and lets us apply the proc to all the uses of
Icon State in the codebase

Sorry if uh, it's crap. Doin what I can

Oh right nearly forgot, this drops plasma walls down more then a
megabyte, to 1.2mbs. Doesn't really do anything to heirophont walls, the
other use of animations we have, but that's alright, I can make those
into overlays anyhow

LemonInTheDark commented:
Handling for more then one delay size is broken, lookin into it

LemonInTheDark committed:
Fuck, ensures delay index always increments

LemonInTheDark commented:
Solved, I was just an idiot

LemonInTheDark commented:
I realize this might work better in the icon crate but I wanna use it now :( (If it's not garbage I can pr it there if you'd like)

LemonInTheDark committed:
removes seemingly unneeded clone()s (I need to read on these more)

actioninja committed:
Merge branch 'master' into dedupe

actioninja committed:
formatting pass

actioninja commented on hypnagogic_core/src/util/icon_ops.rs:
```diff
@@ -0,0 +1,45 @@
+    }
+    new_delays.push(current_delay);
+
+    icon_state.frames = frame_count;
+    icon_state.images = new_images;
+    icon_state.delay = Some(new_delays);
+    return icon_state;
```
More idiomatic to just create a new struct with struct update syntax here instead of modifying the existing one. It's already copied in at the top, no performance difference
Also trailing return

```suggestion
    IconState {
        frames: frame_count,
        images: new_images,
        delay: new_delays,
        ..icon_state
    }
```

actioninja commented on hypnagogic_core/src/util/icon_ops.rs:
```diff
@@ -0,0 +1,45 @@
+    if icon_state.frames <= 1 {
+        return icon_state;
+    }
+    let current_delays = match icon_state.delay {
+        Some(delay) => delay,
+        None => return icon_state,
+    };
```
Good candidate for the fairly recently stablized let else
```suggestion
    let Some(current_delays) = icon_state.delay else {
        return icon_state;
    };
```

actioninja commented on hypnagogic_core/src/util/icon_ops.rs:
```diff
@@ -0,0 +1,45 @@
+        None => return icon_state,
+    };
+
+    // As we walk through the frames in this icon state, we're going to keep track of the ones that
+    // Are duplicates, and "dedupe" them by simply adding extra frame delay and removing the extra frame
+    let mut delay_index = 0;
+    let mut current_delay = current_delays[delay_index];
+    let mut frame_count = icon_state.frames;
+    let mut previous_bytes = None;
+    // List of concrete delays. We'll push to this once we're sure we're happy with the current value
+    let mut new_delays = Vec::new();
+    let mut new_images = Vec::new();
+    for image in icon_state.images {
+        let image_bytes = image.clone().into_bytes();
+        if let Some(previous) = previous_bytes {
+            if previous == image_bytes {
+                previous_bytes = Some(previous);
+                frame_count -= 1;
+                delay_index += 1;
+                current_delay += current_delays[delay_index];
+                continue;
+            }
+            new_delays.push(current_delay);
+            delay_index += 1;
+            current_delay = current_delays[delay_index];
+        }
+        previous_bytes = Some(image_bytes);
+        new_images.push(image);
+    }
+    new_delays.push(current_delay);
```
this is a wonk way of doing this in rust, but it works and explaining more of why would be difficult. cloning each time around is a bit of a hack, ideally you would hold a reference instead. Lifetimes get dicey in loops, so I get why you didn't.

Maybe doing it via a `fold` iterator might be easier with lifetimes? could be tricky if you're not already familiar with complicated iterator usage.

LemonInTheDark replied:
I'm not sure about lifetimes. I did just realize I don't think I needed to convert the image into bytes, since it implements PartialEq

Gonna see about converting this to fold() for fun and no profit tho

actioninja reviewed (changes requested):
some of my review comments disappeared when I ran format

LemonInTheDark committed:
Converts deduping to fold() rather then a for loop

Easier to read.
Also uh, I never neaded to clone() the frame list, since uh, frames
implement partialEq, and since they're just lists of bytes it's
basically free.

LemonInTheDark committed:
trailing return

actioninja commented:
CI is failing anyways, whatever I forgot to merge this forever ago

LemonInTheDark commented:
Pogged
